src/components/ha-template.ts

```diff
@@ -0,0 +1,63 @@
+import type { PropertyValues } from "lit";
+import { LitElement } from "lit";
+import { customElement, property } from "lit/decorators";
+import nunjucks, { type Template as NunjucksTemplate } from "nunjucks";
+import type { HomeAssistant } from "../types";
+
+@customElement("ha-template")
+export class HaTemplate extends LitElement {
+  @property({ attribute: false }) public hass?: HomeAssistant;
+
+  @property({ attribute: false }) public content!: string;
+
+  private _template: NunjucksTemplate | undefined;
+
+  private _accessedEntityIds: string[] = [];
+
+  protected shouldUpdate(changedProperties: PropertyValues): boolean {
+    return (
+      super.shouldUpdate(changedProperties) &&
+      changedProperties.has("hass") &&
+      (!this._accessedEntityIds.length ||
+        changedProperties.has("content") ||
+        this._accessedEntityIds.some(
+          (id) =>
+            this.hass?.states[id].state !==
+            changedProperties.get("hass")?.states[id].state
+        ))
+    );
+  }
+
+  protected willUpdate(changedProperties: PropertyValues): void {
+    this._accessedEntityIds = [];
+    if (changedProperties.has("content")) {
+      this._template = nunjucks.compile(this.content);
+    }
+  }
+
+  protected render() {
+    try {
+      return this._template!.render({
+        hass: this.hass,
+        states: (id: string) => {
+          this._accessedEntityIds.push(id);
```

Comment:
If content changes, we should clear `accessedEntithIds` I guess?

Reply:
Also, we are gonna push the entity id every time we render, would be better to make it a set probably to avoid duplicates?

Reply:
It's cleared in `willUpdate`